src/metabase/data_apps/models.clj

```diff
@@ -0,0 +1,117 @@
+;;                                   :model/DataAppDefinition                                     ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(t2/deftransforms :model/DataAppDefinition
+  {:config mi/transform-json})
+
+(mr/def ::AppDefinitionConfig
+  [:maybe [:map {:closed true}
+           [:version pos-int?]]])
+
+(defn- validate-app-definition
+  "Validate an AppDefinition."
+  [app-definition]
+  (mu/validate-throw ::AppDefinitionConfig (:config app-definition)))
+
+(t2/define-before-insert :model/DataAppDefinition
+  [instance]
+  (validate-app-definition instance)
+  instance)
+
+(t2/define-before-update :model/DataAppDefinition
+  [instance]
+  (throw (ex-info "AppDefinition is append-only and cannot be updated"
+                  {:status-code 400
+                   :changes     (t2/changes instance)})))
+
+(defn- next-revision-number
+  "Get the next revision number for a given app."
+  [app-id]
+  (inc (or (:max (t2/select-one [:model/DataAppDefinition [:%max.revision_number :max]] :app_id app-id))
+           0)))
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                    :model/DataAppRelease                                       ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(t2/define-before-insert :model/DataAppRelease
+  [instance]
+  (merge {:published_at :%now} instance))
+
+(t2/define-before-update :model/DataAppRelease
+  [instance]
+  (let [changed-keys (-> instance t2/changes keys set)]
+    (when-not (= #{:retracted} changed-keys)
+      (throw (ex-info (format "AppRelease is append-only. Only 'retracted' field can be updated, but got: %s"
+                              changed-keys)
+                      {:status-code 400}))))
+  instance)
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                       Serialization                                            ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(comment
+  #_todo)
+
+;;------------------------------------------------------------------------------------------------;;
+;;                                        Public APIs                                             ;;
+;;------------------------------------------------------------------------------------------------;;
+
+(defn new-definition!
+  "Create a new definition for an existing app."
+  [app-id definition]
+  (t2/insert-returning-instance! :model/DataAppDefinition
+                                 (merge definition
+                                        {:app_id          app-id
+                                         :revision_number (next-revision-number app-id)})))
```

Comment:
This is still racey - let's rather use a custom SQL expression to calculate this version on-the-fly as part of the statement.